rpc/v8/subscriptions.go

```diff
@@ -469,76 +513,64 @@ func (h *Handler) SubscribePendingTxs(ctx context.Context, getDetails *bool, sen
 		return 0, rpccore.ErrTooManyAddressesInFilter
 	}
 
-	id := h.idgen()
-	subscriptionCtx, subscriptionCtxCancel := context.WithCancel(ctx)
-	sub := &subscription{
-		cancel: subscriptionCtxCancel,
-		conn:   w,
-	}
-	h.subscriptions.Store(id, sub)
-
-	pendingSub := h.pendingBlock.Subscribe()
-	sub.wg.Go(func() {
-		defer func() {
-			h.unsubscribe(sub, id)
-			pendingSub.Unsubscribe()
-		}()
-
-		h.processPendingTxs(subscriptionCtx, getDetails != nil && *getDetails, senderAddr, pendingSub, w, id)
-	})
+	var (
+		sentTxHashes   map[felt.Felt]struct{}
+		lastParentHash *felt.Felt
+	)
 
-	return SubscriptionID(id), nil
-}
+	return h.subscribe(ctx, w, subscriber{
+		onPending: func(ctx context.Context, id uint64, pending *core.Block) error {
+			if lastParentHash == nil || !pending.ParentHash.Equal(lastParentHash) {
+				sentTxHashes = make(map[felt.Felt]struct{})
+				lastParentHash = pending.ParentHash
+			}
 
-func (h *Handler) processPendingTxs(ctx context.Context, getDetails bool, senderAddr []felt.Felt,
-	pendingSub *feed.Subscription[*core.Block], w jsonrpc.Conn, id uint64,
-) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case pendingBlock := <-pendingSub.Recv():
-			filteredTxs := h.filterTxs(pendingBlock.Transactions, getDetails, senderAddr)
+			filteredTxs := h.filterTxs(pending.Transactions, getDetails != nil && *getDetails, senderAddr, sentTxHashes)
 			for _, filteredTxn := range filteredTxs {
 				if err := sendPendingTxs(w, filteredTxn, id); err != nil {
-					h.log.Warnw("Error sending pending transactions", "err", err)
-					return
+					return err
 				}
 			}
-		}
-	}
+			return nil
+		},
+	})
+}
+
+func toFullTx(txn core.Transaction) any {
+	return AdaptTransaction(txn)
+}
+
+func toHash(txn core.Transaction) any {
+	return txn.Hash()
 }
 
 // filterTxs filters the transactions based on the getDetails flag.
 // If getDetails is true, response will contain the transaction details.
 // If getDetails is false, response will only contain the transaction hashes.
-func (h *Handler) filterTxs(pendingTxs []core.Transaction, getDetails bool, senderAddr []felt.Felt) []any {
+func (h *Handler) filterTxs(
+	pendingTxs []core.Transaction,
+	getDetails bool,
+	senderAddr []felt.Felt,
+	sentTxHashes map[felt.Felt]struct{},
+) []any {
+	var toResult func(txn core.Transaction) any
 	if getDetails {
-		return h.filterTxDetails(pendingTxs, senderAddr)
+		toResult = toFullTx
+	} else {
+		toResult = toHash
 	}
-	return h.filterTxHashes(pendingTxs, senderAddr)
-}
-
-func (h *Handler) filterTxDetails(pendingTxs []core.Transaction, senderAddr []felt.Felt) []any {
 	filteredTxs := make([]any, 0, len(pendingTxs))
```

Comment:
Since we already now the length of `filteredTxs` let's use the real len and make assignments of the form `filteredTxs[i] = smth`, which it has been proven to be less costly. I am aware it was not made in this PR but I think it is good time as ever to change it